src/units/abilities.cpp

```diff
@@ -2048,7 +2059,15 @@ effect::effect(const unit_ability_list& list, int def, const_attack_ptr att, boo
 		effect_list_.push_back(val.second);
 	}
 
-	composite_value_ = static_cast<int>((value_set + addition) * multiplier / divisor);
+	if(max_value && min_value) {
+		composite_value_ = std::clamp(*min_value, *max_value, static_cast<int>((value_set + addition) * multiplier / divisor));
+	} else if(max_value) {
+		composite_value_ = std::min(*max_value, static_cast<int>((value_set + addition) * multiplier / divisor));
+	} else if(min_value) {
+		composite_value_ = std::max(*min_value, static_cast<int>((value_set + addition) * multiplier / divisor));
+	} else {
+		composite_value_ = static_cast<int>((value_set + addition) * multiplier / divisor);
```

Comment:
All four lines seem to do this part, so instead of making this conditional it could be:
```
	composite_value_ = static_cast<int>((value_set + addition) * multiplier / divisor);
	if(max_value && min_value) {
		composite_value_ = std::clamp(*min_value, *max_value, composite_value_);
	} else if(max_value) {
		composite_value_ = std::min(*max_value, composite_value_);
	} else if(min_value) {
		composite_value_ = std::max(*min_value, composite_value_);
	}
```